### PR TITLE
[WIP] Choice equivalent for PyTorch

### PR DESCRIPTION
**Pull request by LeviViana**
Related to #16897.

I'm having a hard time with `.../pytorch/build/lib/libcaffe2.so: undefined reference to 'at::native::choice(at::Tensor const&, at::Tensor const&, bool, long)'`. I tried many workarounds and read some recent PR's trying to figure out how to solve this, but the main problem is that the trial-and-error cycle is too long due to the build time (30 min), so I think I'm gonna need some help in integrating it into PyTorch.

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -4981,6 +4981,11 @@
     CPU: upsample_nearest3d_backward_cpu
     CUDA: upsample_nearest3d_backward_cuda
 
+- func: choice(Tensor& input, Tensor& weights, bool replacement, int k) -> Tensor
```
Comment (soumith) [inline aten/src/ATen/native/native_functions.yaml]: not Tensor&, just Tensor

**Review by soumith — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/TensorFactories.cpp [hunk 2]
```diff
@@ -727,5 +727,133 @@ Tensor tensor_cuda(ArrayRef<T> values, const TensorOptions& options) {
...
+    Tensor cdf = weights.cumsum(0);
+    cdf /= cdf[-1];
+    samples = (uniform_samples.unsqueeze(1) > cdf.unsqueeze(0)).sum(1);
+  }
+
+  return x.index_select(0, samples);
+}
+
+Tensor choice_cpu(
+  Tensor& input,
```
Comment (ezyang) [inline aten/src/ATen/native/TensorFactories.cpp]: These arguments should be `const Tensor&` (and fix @soumith's comment too)

**Comment by ezyang** [comment]
@LeviViana Meta advice: the translation to c++ is documented in the README in https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native

Also, we generate prototypes for all native functions in `build/aten/src/ATen/NativeFunctions.h` which you can use to find the correct spelling.

**Review by fmassa — CHANGES_REQUESTED** [review]
Could you add some tests?

**Comment by LeviViana** [comment]
The reproducibility tests are not passing for the moment. I'm trying to get some inspiration on how `randperm` is implemented to make it pass. Any tips on how to do it will be much appreciated.

File: aten/src/ATen/native/TensorFactories.cpp [hunk 3]
```diff
@@ -727,5 +727,184 @@ Tensor tensor_cuda(ArrayRef<T> values, const TensorOptions& options) {
...
+
+  int n = x.size(0);
+
+  AT_CHECK(
+    n >= k,
+    "Cannot take a larger sample than population when 'replace=False'"
+  );
+
+  auto options = x.options().dtype(at::kLong);
+  THGenerator* generator = THGenerator_new();
```
Comment (fmassa) [inline aten/src/ATen/native/TensorFactories.cpp]: This creates a new random number generator. You should ideally get the generator from somewhere else, or else you won't be able to set a random seed. I'm not sure what's the best option here now, given that the RNG is being completely refactored in https://github.com/pytorch/pytorch/pull/16604

File: aten/src/ATen/native/cuda/TensorFactories.cu [hunk 4]
```diff
@@ -413,4 +419,179 @@ Tensor triu_indices_cuda(
...
+      gen_states
+    );
+
+    generate_reservoir<<<1, 1>>>(
+      indices_n.data<int64_t>(),
+      samples.data<int64_t>(),
+      nb_iterations,
+      split
+    );
+
```
Comment (fmassa) [inline aten/src/ATen/native/cuda/TensorFactories.cu]: Can you add a `AT_CUDA_CHECK(cudaGetLastError());` here and after you called the kernels in the other branch?



AT_CUDA_CHECK(cudaGetLastError());
707

` here and in the other branch?

(note: there is probably a different way of doing this with ATen now, I vaguely remember that something has changed :-)

**Comment by LeviViana** [comment]
Sorry, I messed up (early rebase). I'm trying to fix it.

**Comment by LeviViana** [comment]
I'm closing this PR and opening a new one (cleaner). Sorry for this mess.

**Pull request closed without merging**